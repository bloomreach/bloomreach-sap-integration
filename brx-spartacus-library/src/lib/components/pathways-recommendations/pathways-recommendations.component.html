<div *ngIf="errorMessage" class="alert alert-danger" role="alert">
  {{ errorMessage }}
</div>

<div *ngIf="!errorMessage" class="container">
  <div class="row">
    <div class="col-12 col-lg-12">
      <div class="pathways-and-recommendations mw-container mx-auto">
        <h4 class="mb-4 title">{{ title }}</h4>
        <cx-carousel
          [items]="(products$ | async) || getempty()"
          [template]="carouselItem"
          itemWidth="285px"
        >
        </cx-carousel>

        <ng-template #carouselItem let-item="item">
          <a
            tabindex="0"
            routerLink="/product/{{ item.code }}"
            class="link-container text-decoration-none text-reset"
          >
            <div class="img-container">
              <cx-media
                *ngIf="item.images?.PRIMARY"
                [container]="item.images?.PRIMARY"
                format="product"
              ></cx-media>
            </div>

            <div *ngIf="showTitle" class="name" [title]="item.name">
              {{ item.name }}
            </div>

            <div *ngIf="showPid" class="text-muted">
              Product No. <br /><span class="text-primary ml-1">{{
                item.code
              }}</span>
            </div>

            <div *ngIf="showDescription" class="description text-muted">
              {{ item.description }}
            </div>

            <div *ngIf="showPrice" class="price">
              {{ item.price?.formattedValue }}
            </div>
          </a>
        </ng-template>
      </div>
    </div>
  </div>
</div>
