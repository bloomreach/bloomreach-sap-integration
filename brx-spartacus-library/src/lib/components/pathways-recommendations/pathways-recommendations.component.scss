div {
    margin-bottom: 10px;
}

a.link-container {
    padding: 5px 40px;
    display: block;
    text-align: center; 
    outline: none;
}

.img-container {
  display: flex;
  justify-content: center;
  margin-bottom: 16px;
}

.title {
    font-weight: bold;
    margin: 10px 0 0 20px;
    font-size: 1.5em;
}

.name {
    font-size: var(--cx-font-size, 1rem);
    font-weight: var(--cx-font-weight-bold);
    line-height: var(--cx-line-height, 1.2222222222);
    text-align: center;
    margin: 0 0 3px 0;
    color: var(--cx-color-dark);
    text-decoration: none;
    overflow: hidden;
    text-overflow: ellipsis;
    box-orient: vertical;
    line-clamp: 2;
    max-width: 100%;
    height: 40px;
    display: block;
    &:hover {
        color: var(--cx-color-primary);;
    }
}

.price {
    font-size: var(--cx-font-size, 1.375rem);
    font-weight: var(--cx-font-weight-semi);
    line-height: var(--cx-line-height, 1.2222222222);
    text-align: center;
    margin: 1rem 0;
    color: #212738;
}

.description {
  margin-bottom: 12px;
  display: block; 
  display: -webkit-box;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;
  line-height: 20px;
  max-height: 60px;
  overflow: hidden;
}