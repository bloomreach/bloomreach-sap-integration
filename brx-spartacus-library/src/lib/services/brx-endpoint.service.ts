/*
 * Copyright 2020-2021 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Inject, Injectable } from '@angular/core';
import { PaginationModel } from '@spartacus/core';
import { BrxSearchConfig } from '../models/brx-search-config.model';
import { configType } from '../models/brx-config-type.model';
import { getCookie } from '../utils/get-cookie';

@Injectable({
  providedIn: 'root',
})
export class BrxEndpointService {
  
  // Get Values Form Env
  envConfig!: configType;

  // Default parameters
  DEFAULT_PARAMS = {
    request_type: 'search',
    fl: [
      'pid',
      'title',
      'brand',
      'price',
      'sale_price',
      'promotions',
      'thumb_image',
      'sku_thumb_images',
      'sku_swatch_images',
      'sku_color_group',
      'url',
      'price_range',
      'sale_price_range',
      'description',
      'is_live',
      'score',
    ].join(','),
    realm: 'prod',
    facet: 'true',
  };

  rows = 12; // default page size as per Spartacus

  sort = '';

  currentPage = 0;

  // Still to configure
  viewId = '';

  catalogViews = '';

  constructor() {}

  buildSearchUrl(query = '', searchConfig: BrxSearchConfig = {}, facets?: string[]): string {
    // console.log('[BrxEndpointService.buildSearchUrl.query]: ', query);
    // console.log('[BrxEndpointService.buildSearchUrl.searchConfig]: ', searchConfig);
    // console.log('[BrxEndpointService.buildSearchUrl.facets]: ', facets);
    const { currentPage, sort, type } = searchConfig;
    const { accountId, domainKey, authKey, smEndPoint } = this.envConfig;
    this.sort = sort || '';
    if (this.sort === 'relevance') {
      this.sort = '';
    }
    const brUid = getCookie(document.cookie, '_br_uid_2'); // Need to grab from Cookie _br_uid_2
    const refUrl = document.referrer ?? '';
    this.currentPage = currentPage ? +currentPage : 0;
    this.rows = searchConfig.pageSize || 12;
    const start = this.currentPage * this.rows;
    let searchType = 'keyword';
    if (type === 'category') {
      searchType = 'category';
    }

    const params = new URLSearchParams({
      account_id: accountId,
      domain_key: domainKey,
      auth_key: authKey,
      view_id: '',
      request_id: Date.now().toString(),
      _br_uid_2: brUid,
      url: window.location.href,
      ref_url: refUrl,
      ...this.DEFAULT_PARAMS,
      search_type: searchType,
      rows: this.rows.toString(),
      start: start.toString(),
      q: query,
      sort: this.sort,
    });

    if (facets?.length) {
      const facetFieldsFilters = new Map<string, string[]>();
      for (let i = 0; i < facets.length; i += 2) {
        const id = facets[i];
        const value = facets[i + 1];
        if (!facetFieldsFilters.has(id)) {
          facetFieldsFilters.set(id, []);
        }
        facetFieldsFilters.get(id)?.push(value);
      }
      this.appendFacetFilterParams(params, 'fq', facetFieldsFilters);
    }

    return `${smEndPoint}v1/core/?${params.toString()}`;
  }

  getPaginationDetails(total: number): PaginationModel {
    return {
      currentPage: this.currentPage,
      pageSize: this.rows,
      sort: this.sort,
      totalPages: Math.ceil(total / this.rows),
      totalResults: total,
    };
  }

  buildSuggestionUrl(term = ''): string {
    const { accountId, domainKey, authKey, smSuggestionEndPoint } = this.envConfig;
    const brUid = getCookie(document.cookie, '_br_uid_2');
    const refUrl = document.referrer ?? '';
    const params = new URLSearchParams({
      account_id: accountId,
      auth_key: authKey,
      request_id: new Date().getTime().toString(),
      _br_uid_2: brUid,
      url: window.location.href,
      ref_url: refUrl,
      request_type: 'suggest',
      rsp_fmt: 'v2',
      q: term,
    });

    if (this.catalogViews) {
      params.append('catalog_views', this.catalogViews);
      params.append('rsp_fmt', 'v2');
      return `${smSuggestionEndPoint}v2/suggest/?${params}`;
    }

    params.append('domain_key', domainKey);
    return `${smSuggestionEndPoint}v1/suggest/?${params}`;
  }

  buildPathwaysandRecommendationsUrl(
    isPreview: boolean,
    queryParam = '',
    searchConfig: BrxSearchConfig = {},
    widgetType: string,
    widgetId: string,
    categoryId = '',
    pids: string,
  ): string {
    const { accountId, domainKey, smEndPoint } = this.envConfig;
    const DUMMY_BR_UID_2_FOR_PREVIEW = 'uid%3D0000000000000%3Av%3D11.5%3Ats%3D1428617911187%3Ahc%3D55';
    const brUid2 = getCookie(document.cookie, '_br_uid_2') || (isPreview ? DUMMY_BR_UID_2_FOR_PREVIEW : undefined);
    const refUrl = document.referrer ?? '';
    const size = searchConfig.pageSize || 12;
    const params = new URLSearchParams({
      account_id: accountId,
      view_id: '',
      domain_key: domainKey,
      request_id: new Date().getTime().toString(),
      _br_uid_2: brUid2 ?? '',
      item_ids: pids,
      cat_id: categoryId,
      query: queryParam ?? '',
      url: window.location.href,
      ref_url: refUrl,
      rows: size.toString(),
      fields: [
        'pid',
        'title',
        'brand',
        'price',
        'sale_price',
        'promotions',
        'thumb_image',
        'sku_thumb_images',
        'sku_swatch_images',
        'sku_color_group',
        'url',
        'price_range',
        'sale_price_range',
        'description',
        'is_live',
        'score',
      ].join(','),
    });

    // switch (widgetType) {
    //   case 'item':
    //     params.append('item_ids', pids);
    //     break;
    //   case 'category':
    //     params.append('cat_id', categoryId);
    //     break;
    //   case 'keyword':
    //   case 'personalized':
    //     params.append('query', query);
    //     break;
    //   default:
    //     params.append('query', '');
    // }
    return `${smEndPoint}v2/widgets/${widgetType}/${widgetId}?${params}`;
  }

  private appendFacetFilterParams(
    params: URLSearchParams,
    filterParamName: string,
    facetFieldsFilters: Map<string, string[]>,
  ): void {
    // console.log('[BrxEndpointService.appendFacetFilterParams]: facetFieldsFilters=', facetFieldsFilters);
    facetFieldsFilters.forEach((values, id) => {
      if (values.length > 0) {
        const filterValues = values
          .map((value) => {
            const escaped = value.replace(/"/g, '\\"');
            return `"${escaped}"`;
          })
          .join(' OR ');
        params.append(filterParamName, `${id}:${filterValues}`);
      }
    });
  }
}
