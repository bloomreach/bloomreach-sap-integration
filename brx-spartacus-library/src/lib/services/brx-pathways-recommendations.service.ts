import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Inject, Injectable } from '@angular/core';
import { ConverterService, PRODUCT_NORMALIZER } from '@spartacus/core';
import { BehaviorSubject } from 'rxjs';
import { map } from 'rxjs/operators';
import { BrxSearchConfig } from '../models/brx-search-config.model';
import { ConfigType } from '../models/brx-config-type.model';
import { BrxEndpointService } from './brx-endpoint.service';

@Injectable({
  providedIn: 'root',
})
export class BrxPathwaysRecommendations {
  constructor(
    protected http: HttpClient,
    protected brxEndpoints: BrxEndpointService,
    protected converter: ConverterService,
  ) {}

  productsSubject = new BehaviorSubject<any>([]);

  getproducts(
    isPreview: boolean,
    query = '',
    searchConfig: BrxSearchConfig,
    widgetType: string,
    widgetid: string,
    categoryid: string,
    pids: string,
  ) {
    const { authKey } = this.brxEndpoints.envConfig;
    const requestOptions = {
      headers: new HttpHeaders({ 'auth-key': authKey }),
    };
    const req = this.http
      .get(
        this.brxEndpoints.buildPathwaysandRecommendationsUrl(
          isPreview,
          query,
          searchConfig,
          widgetType,
          widgetid,
          categoryid,
          pids,
        ),
        requestOptions,
      )
      .pipe(
        map((data: any) => {
          return data?.response?.docs.map((product: any) => this.converter.convert(product, PRODUCT_NORMALIZER)) ?? {};
        }),
      );

    req.subscribe((data) => {
      this.productsSubject.next(data);
    });
  }
}
