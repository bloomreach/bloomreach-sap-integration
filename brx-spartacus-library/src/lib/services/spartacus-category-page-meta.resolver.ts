/*
 * Copyright 2020-2021 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Injectable } from '@angular/core';
import {
  BasePageMetaResolver,
  Breadcrumb,
  BreadcrumbMeta,
  CategoryPageMetaResolver,
  CmsService,
  Page,
  ProductSearchPage,
  TranslationService,
} from '@spartacus/core';
import { Observable } from 'rxjs';
import { filter, switchMap } from 'rxjs/operators';
import { SpartacusProductSearchService } from '.';

@Injectable({
  providedIn: 'root',
})
export class SpartacusCategoryPageMetaResolver extends CategoryPageMetaResolver {
  constructor(
    protected productSearchService: SpartacusProductSearchService,
    protected cms: CmsService,
    protected translation: TranslationService,
    protected basePageMetaResolver: BasePageMetaResolver,
  ) {
    super(productSearchService, cms, translation, basePageMetaResolver);
  }

  resolveTitle(): Observable<string> {
    return (<Observable<ProductSearchPage>>this.searchPage$).pipe(
      filter((page: ProductSearchPage) => !!page.pagination),
      switchMap((p: ProductSearchPage) =>
        this.translation.translate('pageMetaResolver.category.title', {
          count: p.pagination?.totalResults,
          query: p.breadcrumbs?.reduce<Breadcrumb | undefined>(
            (prev, curr) => (curr.facetCode === 'category' || curr.facetCode === 'allCategories' ? curr : prev),
            undefined,
          )?.facetValueName,
        }),
      ),
    );
  }

  /**
   * Modified from /product/services/category-page-meta.resolver.ts
   * of SAP Spartacus Core Library project.
   *
   * Resolves breadcrumbs for the Category Page. The breadcrumbs are driven by
   * a static home page crumb and a crumb for each category.
   */
  protected resolveBreadcrumbData(page: ProductSearchPage, label: string): BreadcrumbMeta[] {
    const breadcrumbs: BreadcrumbMeta[] = [];
    breadcrumbs.push({ label, link: '/' });

    (page.breadcrumbs ?? []).forEach((br) => {
      if (br.facetValueName) {
        if (br.facetCode === 'category' || br.facetCode === 'allCategories') {
          breadcrumbs.push({
            label: br.facetValueName,
            link: `/category/${br.facetValueCode}`,
          });
        }
      }
    });
    return breadcrumbs;
  }

  /**
   * Ensure this resolver is always getting picked over the default one.
   */
  getPriority(page: Page): number {
    return this.getScore(page) + 1;
  }
}
