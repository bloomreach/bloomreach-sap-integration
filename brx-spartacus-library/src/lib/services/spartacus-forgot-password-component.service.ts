/*
 * Copyright 2020-2021 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Injectable } from '@angular/core';
import { AuthConfigService, GlobalMessageService, RoutingService } from '@spartacus/core';
import { ForgotPasswordComponentService } from '@spartacus/user/profile/components';
import { UserPasswordFacade } from '@spartacus/user/profile/root';

@Injectable({
  providedIn: 'root',
})
export class SpartacusForgotPasswordComponentService extends ForgotPasswordComponentService {
  constructor(
    protected userPasswordService: UserPasswordFacade,
    protected routingService: RoutingService,
    protected authConfigService: AuthConfigService,
    protected globalMessage: GlobalMessageService
  ) {
    super(userPasswordService, routingService, authConfigService, globalMessage)
  }
}