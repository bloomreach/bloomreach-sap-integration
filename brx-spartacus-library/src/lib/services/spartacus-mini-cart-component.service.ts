/*
 * Copyright 2024 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Injectable } from '@angular/core';
import { AuthService, EventService, SiteContextParamsService, StatePersistenceService } from '@spartacus/core';
import { ActiveCartFacade } from '@spartacus/cart/base/root';
import { MiniCartComponentService } from '@spartacus/cart/base/components/mini-cart';
import { Observable, of } from 'rxjs';
import { map, startWith, switchMap } from 'rxjs/operators';

@Injectable({
  providedIn: 'root',
})
export class SpartacusMiniCartComponentService extends MiniCartComponentService {
  constructor(
    protected activeCartFacade: ActiveCartFacade,
    protected authService: AuthService,
    protected statePersistenceService: StatePersistenceService,
    protected siteContextParamsService: SiteContextParamsService,
    protected eventService: EventService,
  ) {
    super(activeCartFacade, authService, statePersistenceService, siteContextParamsService, eventService);
  }

  getQuantity(): Observable<number> {
    return this.activeCartRequired().pipe(
      switchMap((activeCartRequired) => {
        if (activeCartRequired) {
          return this.activeCartFacade.getActive().pipe(
            startWith({ totalUnitCount: 0, deliveryItemsQuantity: 0 }),
            map((cart) => cart.totalUnitCount || cart.deliveryItemsQuantity || 0),
          );
        }
        return of(0);
      }),
    ) as Observable<number>;
  }
}
