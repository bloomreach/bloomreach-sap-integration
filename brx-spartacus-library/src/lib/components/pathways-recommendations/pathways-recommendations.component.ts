import { Component, Input, OnInit } from '@angular/core';
import { ContainerItem, getContainerItemContent, Page } from '@bloomreach/spa-sdk';

import { Observable, of } from 'rxjs';

import { map } from 'rxjs/operators';
import { BrxPathwaysRecommendations } from '../../services';
import { PathwaysRecommendationsCompound, PathwaysRecommendationsParameters } from '../../models/common-types.model';

@Component({
  selector: 'brx-pathways-recommendations',
  templateUrl: './pathways-recommendations.component.html',
  styleUrls: ['./pathways-recommendations.component.scss'],
})
export class PathwaysRecommendationsComponent implements OnInit {
  @Input() component!: ContainerItem;

  @Input() page!: Page;

  title = '';

  errorMessage: string | undefined;

  interval: number | undefined;

  limit: number | undefined;

  showDescription = false;

  showPid = false;

  showPrice = false;

  showTitle = false;

  products$: Observable<Observable<any>[]> | undefined;

  constructor(private brxpathwaysrecommendationsservice: BrxPathwaysRecommendations) {}

  ngOnInit(): void {
    this.renderWrappedPathways();
    this.products$ = this.brxpathwaysrecommendationsservice.productsSubject.pipe(
      map((item: any) => Object.values(item).map((val: any) => of(val))),
    );
  }

  getempty(): Observable<any>[] {
    return [of()];
  }

  renderWrappedPathways(): void {
    // eslint-disable-next-line max-len
    const { interval, limit, title, showDescription, showPid, showPrice, showTitle } =
      this.component.getParameters<PathwaysRecommendationsParameters>();

    this.title = title || '';
    this.interval = interval || 0;
    this.limit = limit || 4;
    this.showDescription = showDescription;
    this.showPid = showPid;
    this.showPrice = showPrice;
    this.showTitle = showTitle;

    const { categoryCompound, keyword, productCompound, widgetCompound } =
      getContainerItemContent<PathwaysRecommendationsCompound>(this.component, this.page) ?? {};

    const { categoryid } = categoryCompound ?? {};

    const pids = productCompound?.map(({ productid }) => {
      const [, id, code] = productid.match(/id=([\w\d._=-]+[\w\d=]?)?;code=([\w\d._=/-]+[\w\d=]?)?/i) ?? [];
      return code ?? id;
    });
    const { widgetid: widgetId = '', widgetalgo: widgetAlgo } = widgetCompound ?? {};
    const widgetType = widgetAlgo?.selectionValues?.[0].key.split('.')[0] ?? '';

    if ((widgetId ?? 'undefined') !== 'undefined' && widgetType) {
      switch (widgetType) {
        case 'item':
          this.errorMessage = !pids ? 'Widget configured incorrectly: please add Product IDs' : undefined;
          break;
        case 'category':
          this.errorMessage = !categoryid ? 'Widget configured incorrectly: please add a Category ID' : undefined;
          break;
        case 'keyword':
        case 'personalized':
          this.errorMessage = !keyword ? 'Widget configured incorrectly: please add a Keyword' : undefined;
          break;
        default:
          this.errorMessage = undefined;
      }
    } else {
      this.errorMessage = 'Please configure Widget ID and Widget Type first';
    }

    this.brxpathwaysrecommendationsservice.getproducts(
      this.page.isPreview(),
      keyword,
      {
        pageSize: this.limit,
      },
      widgetType,
      widgetId,
      categoryid || '',
      pids?.join(',') ?? '',
    );
  }
}
