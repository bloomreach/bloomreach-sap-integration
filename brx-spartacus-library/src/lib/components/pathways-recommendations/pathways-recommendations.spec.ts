import { ComponentFixture, TestBed } from '@angular/core/testing';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { RouterModule } from '@angular/router';
import { APP_BASE_HREF } from '@angular/common';
import { Page } from '@bloomreach/spa-sdk';
import { BehaviorSubject } from 'rxjs';
import { PathwaysRecommendationsComponent } from './pathways-recommendations.component';
import { BrxPathwaysRecommendations } from '../../services';
import mockComponent from '../../../__mocks__/container-item';
import mockPage from '../../../__mocks__/page';

describe('PathwaysRecommendationsComponent', () => {
  let component: PathwaysRecommendationsComponent;
  let componentEl: HTMLElement;
  let fixture: ComponentFixture<PathwaysRecommendationsComponent>;

  const productsSubject = new BehaviorSubject([]);

  const widgetId = 'widgetid';
  const keyword = 'keyword';
  const productId = 'productId';
  const productCode = 'productCode';
  const categoryId = 'categoryId';
  const selectionValueKey = 'selectionValueKey';
  const mockPageContent = {
    type: 'componentcontent',
    data: {
      widgetCompound: {
        widgetid: widgetId,
        widgetalgo: {
          selectionValues: [{ key: `${selectionValueKey}.test` }],
        },
      },
      keyword,
      categoryCompound: {
        categoryid: categoryId,
      },
      productCompound: [{ productid: `id=${productId};code=` }, { productid: `id=;code=${productCode}` }],
    },
  };

  const mockBrxPathwaysRecommendations: Partial<BrxPathwaysRecommendations> = {
    getproducts: jest.fn(),
    productsSubject,
  };

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [RouterModule],
      declarations: [PathwaysRecommendationsComponent],
      providers: [
        { provide: BrxPathwaysRecommendations, useValue: mockBrxPathwaysRecommendations },
        { provide: APP_BASE_HREF, useValue: '/' },
      ],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
      teardown: { destroyAfterEach: false },
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(PathwaysRecommendationsComponent);
    component = fixture.componentInstance;
    componentEl = fixture.nativeElement;

    component.component = {
      ...mockComponent,
    };

    component.page = {
      ...mockPage,
      isPreview: jest.fn(() => false),
      getContent: jest.fn(() => mockPageContent),
    } as unknown as Page;

    fixture.detectChanges();
  });

  it('should initialize component correctly', () => {
    jest.spyOn(component, 'renderWrappedPathways');
    productsSubject.next({ product: 'test' } as any);

    component.ngOnInit();

    fixture.detectChanges();

    expect(component.renderWrappedPathways).toHaveBeenCalled();
    expect(fixture.nativeElement).toMatchSnapshot();
  });

  describe('renderWrappedPathways', () => {
    it('should call getproducts', () => {
      component.renderWrappedPathways();

      expect(mockBrxPathwaysRecommendations.getproducts).toHaveBeenCalledWith(
        false,
        keyword,
        { pageSize: 4 },
        selectionValueKey,
        widgetId,
        categoryId,
        `${productId},${productCode}`,
      );
    });

    it('should use default parameters when none are passed', () => {
      jest.spyOn(component.component, 'getParameters').mockReturnValueOnce({});

      component.renderWrappedPathways();

      expect(component.title).toBe('');
      expect(component.interval).toBe(0);
      expect(component.limit).toBe(4);
    });

    it('should show an error message if widgetId or widgetType are missing', () => {
      const mockContentData = { ...mockPageContent };
      mockContentData.data.widgetCompound = {} as any;
      jest.spyOn(component.page, 'getContent').mockReturnValueOnce(mockContentData);

      component.renderWrappedPathways();

      fixture.detectChanges();

      expect(component.errorMessage).toBe('Please configure Widget ID and Widget Type first');
      expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('should show an error message if keyword is missing', () => {
      const mockContentData = { ...mockPageContent };
      const mockWidgetType = 'keyword';
      mockContentData.data.keyword = '';
      mockContentData.data.widgetCompound = {
        widgetid: widgetId,
        widgetalgo: {
          selectionValues: [{ key: `${mockWidgetType}.test` }],
        },
      };
      jest.spyOn(component.page, 'getContent').mockReturnValueOnce(mockContentData);

      component.renderWrappedPathways();

      expect(component.errorMessage).toBe('Widget configured incorrectly: please add a Keyword');
    });

    it('should show an error message if category id is missing', () => {
      const mockContentData = { ...mockPageContent };
      const mockWidgetType = 'category';
      mockContentData.data.categoryCompound.categoryid = '';
      mockContentData.data.widgetCompound = {
        widgetid: widgetId,
        widgetalgo: {
          selectionValues: [{ key: `${mockWidgetType}.test` }],
        },
      };
      jest.spyOn(component.page, 'getContent').mockReturnValueOnce(mockContentData);

      component.renderWrappedPathways();

      expect(component.errorMessage).toBe('Widget configured incorrectly: please add a Category ID');
    });

    it('should show an error message if product ids are missing', () => {
      const mockContentData = { ...mockPageContent };
      const mockWidgetType = 'item';
      mockContentData.data.productCompound = undefined as any;
      mockContentData.data.widgetCompound = {
        widgetid: widgetId,
        widgetalgo: {
          selectionValues: [{ key: `${mockWidgetType}.test` }],
        },
      };
      jest.spyOn(component.page, 'getContent').mockReturnValueOnce(mockContentData);

      component.renderWrappedPathways();

      expect(component.errorMessage).toBe('Widget configured incorrectly: please add Product IDs');
    });
  });
});
