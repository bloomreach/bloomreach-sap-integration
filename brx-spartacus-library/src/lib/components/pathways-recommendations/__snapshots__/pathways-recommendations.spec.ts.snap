// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PathwaysRecommendationsComponent renderWrappedPathways should show an error message if widgetId or widgetType are missing 1`] = `
<div
  id="root3"
>
  
  <div
    class="alert alert-danger"
    role="alert"
  >
     Please configure Widget ID and Widget Type first

  </div>
  
</div>
`;

exports[`PathwaysRecommendationsComponent should initialize component correctly 1`] = `
<div
  id="root0"
>
  
  
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col-12 col-lg-12"
      >
        <div
          class="pathways-and-recommendations mw-container mx-auto"
        >
          <h4
            class="mb-4 title"
          >
            test
          </h4>
          <cx-carousel
            itemwidth="285px"
          />
          
        </div>
      </div>
    </div>
  </div>
</div>
`;
